Landing Page Type

kave opened the pull request:
## Additions

- New Landing Page type for wagtail

## Changes

- Routed landing page type related molecules and organisms to use `jinja2/v1/_includes/` template locations

## Testing

- `pip install -r requirements/base.txt`
- Create landing page type and fill out content.

## Review

- @kurtw 
- @rosskarchner 
- @jimmynotjim 
- @anselmbradford 
- @KimberlyMunoz 
- @sebworks

kave commented on CHANGELOG.md:
```diff
@@ -91,6 +91,7 @@ Given the `MAJOR.MINOR.PATCH` pattern, here is how we decide to increment:
 - Added Contact snippet and molecule backends
 - Added temporary folder for converted Jinja2 Wagtail field template files
+- Added WP Import Data Contact processor
 - Added templates and CSS for the Adding Sidebar Breakout organism.
```
-_- not sure how this snuck in here

kave commented on cfgov/jinja2/v1/_includes/molecules/hero.html:
```diff
@@ -34,31 +34,65 @@
-{% macro render( settings, options={} ) %}
-
-{% if settings.heading and settings.body and settings.image_url %}
+{% if value %}
```
I choose to edit the templates directly instead of creating a new file because it would seem redundant to maintain two separate copies. this will lead to the `jinja2\v1\wagtail` directory eventually being deleted

sebworks commented on cfgov/v1/__init__.py:
```diff
@@ -15,9 +16,25 @@ def environment(**options):
+
+
+@contextfunction
+def render_stream_child(context, stream_child):
```
Is this function necessary to render Jinja2 templates in Wagtail?

kave replied:
its necessary to specifically to streamblocks with richtext field. without it, it prints those stream fields as raw html instead of rendering them.

sebworks replied:
awesome, thanks

kave replied:
yea i'd like to get rid of this eventually, hopefully they will implement a cleaner solution in the next release

kurtw commented on cfgov/v1/models/molecules.py:
```diff
@@ -79,7 +79,7 @@ class ImageText2575(blocks.StructBlock):
         error_dict = {}
         try:
-            block_data = super(ImageText5050, self).clean(data)
+            block_data = super(ImageText2575, self).clean(data)
```
lol this was probably my handy work.

kurtw commented on cfgov/v1/models/molecules.py:
```diff
@@ -117,16 +117,76 @@ class TextIntroduction(blocks.StructBlock):
+    image_path = blocks.CharBlock(required=False)
+    image_alt = blocks.CharBlock(required=False)
+
+    background_color = blocks.CharBlock(max_length=100, required=False)
```
Not sure if this is something we should be handling in the CMS or not. @sebworks @anselmbradford @jimmynotjim @KimberlyMunoz how is background determined in the front end? If there's a manageable set of background colors to use then we can set up a ChoiceField to select from that list if it makes sense to do so. What do yall think?

jimmynotjim replied:
I think that's actually a question for designers. I know there's at least a few colors they wanted to choose from, maybe we make it a dropdown? cc @ajbush @duelj

kave replied:
A dropdown could work

kurtw commented on cfgov/jinja2/v1/_includes/molecules/hero.html:
```diff
@@ -34,31 +34,65 @@
+                <a href="{{ value.link_url }}"
                    class="hero_cta
-                   {{ 'btn' if options.is_button
+                   {{ 'btn' if value.is_button
```
I'm not seeing the button in the designs. @anselmbradford?

jimmynotjim replied:
It's in the [Design Manual Spec](http://stephanieosan.github.io/design-manual/ui-toolkit/heroes.html) as "Call to Action Button"

kurtw replied:
Right, but I'm not seeing that in the Hero specs. It's usually called CTA standalone w/ button right?

jimmynotjim replied:
It's an option to turn the link into a button. Design just needs to update the v1 Hero spec.

kurtw replied:
@ajbush @duelj

kurtw commented on cfgov/v1/models/molecules.py:
```diff
@@ -117,16 +117,76 @@ class TextIntroduction(blocks.StructBlock):
+
+        if block_data['image_path'] and not block_data['image_alt']:
+            error_dict.update({'image_alt': isRequired('Image Alt')})
+
```
I think there should also be a validation error if there's an `image_alt` but not an `image_path`.

kave replied:
True, i will add that in the next PR where i've extracted this logic out into the atoms model file

kurtw replied:
:+1:

jimmynotjim replied:
Hero images should always be background images and thus don't need an `alt` attribute

kurtw replied:
Thanks for mentioning that.

kurtw commented on cfgov/v1/models/molecules.py:
```diff
@@ -117,16 +117,77 @@ class TextIntroduction(blocks.StructBlock):
+    btn_text = blocks.CharBlock(max_length=100, required=True)
+
+    required = blocks.BooleanBlock(required=False)
+    id = blocks.CharBlock(max_length=100, required=False)
```
Possibly remove.

kurtw commented on cfgov/v1/models/molecules.py:
```diff
@@ -117,16 +117,77 @@ class TextIntroduction(blocks.StructBlock):
+    id = blocks.CharBlock(max_length=100, required=False)
+    info = blocks.RichTextBlock(required=False)
+    label = blocks.CharBlock(max_length=100, required=True)
+    type = blocks.CharBlock(max_length=100, required=True)
```
Now I'm wondering if this, and other fields, can be passed in when the block is declared.

For example from the Email org: `form_field = blocks.ListBlock(molecules.FormFieldWithButton(type='email'), icon='mail', required=False)`

kurtw replied:
Via verbal agreement, we're changing this to a ChoiceBlock.

kave replied:
i'll turn this into a chooser block instead with all the available input tag types

kave commented on cfgov/jinja2/v1/landing-page/index.html:
```diff
@@ -14,129 +14,54 @@
+        {% endfor %}
     </div>
     <div class="block">
         <h2>Headline for 2/7 organism</h2>
```
@KimberlyMunoz hey can this heading be removed?

KimberlyMunoz replied:
Yes. We might have to add a way to add a title to a group like this. Should we make an atom macro for that?

jimmynotjim replied:
I think there's supposed to always be a headline over a batch of 25/75 molecules cc @ajbush @duelj

KimberlyMunoz replied:
Do you think it would be okay to make an atom molecule for a group title? Or should we adjust the image-text molecule to take this batching into account?

jimmynotjim replied:
I think an atom that can be used here and anywhere else we have a chunk of molecules is a great solution. cc @ajbush @duelj

jimmynotjim commented on cfgov/v1/models/molecules.py:
```diff
@@ -117,16 +117,77 @@ class TextIntroduction(blocks.StructBlock):
+            error_dict.update({'image': img_err, 'image_path': img_err, 'image_alt': isRequired('Image alt')})
+
+        if block_data['image'] and block_data['image_path']:
+            img_err = ['Please select one method of image rendering']
```
What are the methods here? Hero images will always be background images.

kave replied:
Yea correct, so abstracted this validation logic into the atom instead,so its not Hero specifc to so its reusable and opted to make `alt` not required

KimberlyMunoz commented on cfgov/jinja2/v1/_includes/molecules/form-field-with-button.html:
```diff
@@ -27,24 +27,46 @@
+                {% endif %}
+            </label>
+            <input id="{{ (value.id or value.name) ~
+                    '_' ~ range(1, 100) | random }}"
```
Because of this _random-number, the input `id` doesn't match the label and we're getting an accessibility error. This should just be `<input id="{{ (value.id or value.name)">` and there should be something to prevent an id from being used more than once.

sebworks replied:
Yep, this won't ensure uniqueness. Maybe use a context function which returns `uuid.uuid1().hex`?

KimberlyMunoz replied:
Before we try to find a solution on the front-end, is there something on the back-end that can detect if an ID is being used more than once in a form-field?

I'm not a big fan of using something like uuid, which is a little ~too~ unique and overkill for this situation, especially since our pages are not on HTTPS. But I'm open to it.

kurtw replied:
So these form field blocks will have to have unique id's across the page?

KimberlyMunoz replied:
Yep, and the id should match `<label for="ID">` for accessibility reasons.

kurtw replied:
So do we need separate fields for `id` and `name`? or are they the same info

kave replied:
yea good question @KimberlyMunoz

kave replied:
Resolved by removing name field so only Id is required. Also added id validation to only allow alphanumberic values along with `-` `_`

KimberlyMunoz commented on cfgov/jinja2/v1/_includes/molecules/form-field-with-button.html:
```diff
@@ -27,24 +27,46 @@
+                    {% endif %}
+                </label>
+                <input id="{{ (settings.field.id or settings.field.name) ~
+                        '_' ~ range(1, 100) | random }}"
```
Same issue as above.

duelj commented:
I think I am following the conversation here and from the design end we would like a title for a group of molecules, like the 50/50 text blob, 25/75, etc. I think the atom idea for the title makes sense and is how we were viewing it from the design end as well. Would it help to have an organism for a group of molecules like this with a header or description paragraph?

kave commented:
@duelj So yes this could work. So the molecule would look like this right?
Organism:
- heading 
- body
- molecule plural of the same type

@KimberlyMunoz This way we wouldn't have to edit the molecules themselves

kurtw commented on cfgov/v1/models/molecules.py:
```diff
@@ -12,121 +12,90 @@ def isRequired(field_name):
 class CallToAction(blocks.StructBlock):
+    heading = blocks.CharBlock(max_length=100, required=True)
     slug = blocks.CharBlock(required=True)
-    paragraph = blocks.RichTextBlock()
```
Why was this removed?

kave replied:
to be consistent with the template

kurtw replied:
@duelj @ajbush @jimmynotjim  [The specifications for the CTA](https://github.cfpb.gov/flapjack/Modules-V1/wiki/Standalone-CTA-\(with-button\)) say that this has paragraphs. Did this change or are the templates just not including those specs?

duelj replied:
We will want the option of paragraph text with the standalone CTA molecule. The lorem ipsum in [this image](https://github.cfpb.gov/github-enterprise-assets/0000/0160/0001/5470/69ca1838-6101-11e5-819d-7a9cc6be43e2.png) would represent how we would use it.

kurtw replied:
In the templates, there's a variable called `header`. Could this be what the paragraph text atom was supposed to be? @KimberlyMunoz https://github.com/cfpb/cfgov-refresh/pull/1154/files#diff-3f1f0f338f0a28057bf1dac682d9dca2R31

kave replied:
these field names will be resolved in a separate PR https://gist.github.com/sebworks/5586cfa6fe01090804df

kurtw commented on cfgov/v1/models/atoms.py:
```diff
@@ -1,6 +1,70 @@
 class Hyperlink(blocks.StructBlock):
-    label = blocks.CharBlock(max_length=50)
-    href = blocks.CharBlock(default='/')
+    text = blocks.CharBlock(max_length=50, required=False)
```
Why were these names changed?

kave replied:
to be consistent with the templates

kurtw replied:
These fields were changed here but not in the templates. Either the templates should change or the fields should change.

kave replied:
huh? they are all changed in the molecule templates that refer to links. can you point one out where your seeing the change not made.

kurtw replied:
https://github.com/cfpb/cfgov-refresh/pull/1154/files#diff-3f1f0f338f0a28057bf1dac682d9dca2R43

sebworks replied:
ok, so the problem is that `value.slug` should be `value.text`. I am referring to https://github.com/cfpb/cfgov-refresh/pull/1154/files#diff-3f1f0f338f0a28057bf1dac682d9dca2R36.

kave replied:
thats one, i can make that change

KimberlyMunoz commented on cfgov/jinja2/v1/_includes/molecules/hero.html:
```diff
@@ -34,31 +34,65 @@
+            {% set photo=image(value.image.upload, 'original') %}
+            <div class="hero_image"
+                 style="background-image: url( {{ photo.url }} );">
+        {% else %}
```
When we talked with content earlier last week, they said they would just want the image upload and not the image url option. Since it isn't clear that you have to put in one or the other until you submit, can we remove the second option? @sarahsimpson09 

![image](https://cloud.githubusercontent.com/assets/1860176/11283981/64acf9fe-8ed5-11e5-9de2-1cc8b18e8888.png)

KimberlyMunoz replied:
And we should be able to remove the alt since it's a background image.

kave replied:
done

KimberlyMunoz commented on cfgov/v1/models/molecules.py:
```diff
@@ -12,121 +12,90 @@ def isRequired(field_name):
     body = blocks.RichTextBlock(required=False)
-    link_url = blocks.URLBlock(required=False)
-    link_text = blocks.CharBlock(max_length=100, required=False)
+    link = atoms.Hyperlink()
```
Can we make this link not-required?

![image](https://cloud.githubusercontent.com/assets/1860176/11284060/cb2514dc-8ed5-11e5-830e-979c42bf9983.png)

kave replied:
sure i can do that

KimberlyMunoz commented on cfgov/v1/models/organisms.py:
```diff
@@ -1,4 +1,8 @@
+from . import molecules
 
 
 class Well(blocks.StructBlock):
```
The well right now should just be one rich text field. The macro originally had a template path for demo-ing purposes, but content should just see that rich text field.

![image](https://cloud.githubusercontent.com/assets/1860176/11284312/0bc7b85e-8ed7-11e5-83e0-9db66be9409e.png)

kave replied:
@KimberlyMunoz is this the same for Full Width Text Copy?

KimberlyMunoz replied:
Yep.

KimberlyMunoz commented on cfgov/v1/models/molecules.py:
```diff
@@ -12,121 +12,90 @@ def isRequired(field_name):
-    link_url = blocks.URLBlock(required=False)
-    link_text = blocks.CharBlock(max_length=100, required=False)
+    image = atoms.ImageBasic()
+    link = atoms.Hyperlink()
```
Can we make this link optional?

kave replied:
will do

KimberlyMunoz commented on cfgov/v1/models/molecules.py:
```diff
@@ -12,121 +12,90 @@ def isRequired(field_name):
-    image_alt = blocks.CharBlock(required=False)
-    link_url = blocks.URLBlock(required=False)
-    link_text = blocks.CharBlock(max_length=100, required=False)
+    image = atoms.ImageBasic()
```
Can we drop the url field and leave only the Chooser and the Alt text field?

![image](https://cloud.githubusercontent.com/assets/1860176/11284485/1cbe144a-8ed8-11e5-806a-55bce24070aa.png)

kave replied:
done

KimberlyMunoz commented on cfgov/v1/models/molecules.py:
```diff
@@ -12,121 +12,97 @@ def isRequired(field_name):
+    ], icon='cup', required=True)
+    name = blocks.CharBlock(max_length=100, required=False)
+    placeholder = blocks.CharBlock(max_length=100, required=False)
+    attributes = blocks.CharBlock(max_length=100, required=False)
```
Other than required, which we have a checkbox for, are there any other attributes we need to account for, or can we remove this? @sebworks ?

sebworks replied:
yes, we just need the `required` setting.

kave replied:
its been removed in the latest push and templates have been updated

KimberlyMunoz commented on cfgov/v1/models/molecules.py:
```diff
@@ -12,121 +12,97 @@ def isRequired(field_name):
+
+    required = blocks.BooleanBlock(required=False)
+    id = blocks.CharBlock(max_length=100, required=False)
+    info = blocks.RichTextBlock(required=False)
```
Can we change this label to "disclaimer" so that folks know what block of text that is?

kave replied:
donzo

KimberlyMunoz commented on cfgov/v1/models/molecules.py:
```diff
@@ -12,121 +12,97 @@ def isRequired(field_name):
+        ('url', 'URL'),
+        ('radio', 'Radio'),
+    ], icon='cup', required=True)
+    name = blocks.CharBlock(max_length=100, required=False)
```
Can the name match the ID to reduce the need for two fields? I could easily see this getting changed and breaking things.

kave replied:
@KimberlyMunoz sure but is there a need for both fields? couldn't we just only use 1 on the templates and get rid of the other?

KimberlyMunoz commented on cfgov/v1/models/molecules.py:
```diff
@@ -12,121 +12,97 @@ def isRequired(field_name):
+    id = blocks.CharBlock(max_length=100, required=False)
+    info = blocks.RichTextBlock(required=False)
+    label = blocks.CharBlock(max_length=100, required=True)
+    type = blocks.ChoiceBlock(choices=[
```
When I tested this I couldn't get the dropdown to appear.

KimberlyMunoz replied:
Pulled in the recent changes, and this is good now!

kave replied:
:+1:

KimberlyMunoz commented:
Great changes on the groupings with the title! I assume those changes mean that the order of those organisms/molecules is now fixed. If so, can we put them in this order from the spec?

- 50-50 molecule
- 25-75 molecule
- Grey well
- Half link blob

![image](https://cloud.githubusercontent.com/assets/1860176/11286366/99984c1a-8ee2-11e5-820f-0752efa9514b.png)

KimberlyMunoz commented:
I can't get the grey well organism to appear on the page.

KimberlyMunoz commented:
Can we have the markup for containers around the different molecules only appear if the molecule is in use? In this case, I don't have any molecules other than the half-link blob, but there's lots of extra spacing from `block` and headings. 

![image](https://cloud.githubusercontent.com/assets/1860176/11286777/ef934168-8ee4-11e5-9c13-50882f5685f4.png)

KimberlyMunoz commented:
There's some feedback from Content on changing the names of some of the fields to make it clearer for people entering content. 

- [ ] Change "Is widescreen" to "Use 16:9 image"
- [ ] Change "is button" to "Show link as button"
- [ ] Enable the ability for multiple links with the 50-50 image. (Also an issue for 75-25 text-image)

kave commented:
@KimberlyMunoz 
- reorganized the groupings in the landing page as you mentioned.
- fixed Well
- Resolved content feedback but i need you to verify the styling on the templates since its now rendering multiple links instead of 1

kave commented:
@KimberlyMunoz For your markup for containers question. In what scenario did you see this happen? trying to duplicate locally

KimberlyMunoz commented:
For the container thing, I replicated by creating a group of half-link blobs and getting extra blank space.

The styling on multiple links looks great, but is it possible to make the multiple links optional? There's going to be circumstances like on the `/the-bureau/` page where links won't be needed.

I still can't get the grey well to appear.

tnferrell commented:
The only things standing out to me are...

It might be nice to have that multiple link feature you added to the Half-Width Blob on the 25-75 Text Image Molecule and 50-50 Image Text. I just imagine someday we'll run into a use case where that's desired and having the flexibility would be nice. 

These are probably more Wagtail interface questions...but will we be able to format text within these input fields? In the screenshots you sent @KimberlyMunoz, I didn't see any WYSIWYG buttons for bolding, italicizing, etc, etc.  The input fields were also small, like the introduction text field in the text introduction molecule. Do they expand on input or does a scrollbar appear? Just thinking about how easy it'll be to proofread and update in the interface.

KimberlyMunoz commented:
> It might be nice to have that multiple link feature you added to the Half-Width Blob on the 25-75 Text Image Molecule and 50-50 Image Text. I just imagine someday we'll run into a use case where that's desired and having the flexibility would be nice.

Done!

>These are probably more Wagtail interface questions...but will we be able to format text within these input fields?

In the thick fields they appear when you are in those fields.

tnferrell commented:
@KimberlyMunoz So speedy. 

Cool and cool. Do you know if the input fields expand when content is entered or is it a scrollbar situation? Or is this a question for the BEWDs?

KimberlyMunoz commented:
Like this when you click on the field.

![image](https://cloud.githubusercontent.com/assets/1860176/11305290/fd053bf4-8f79-11e5-886c-51bcb08b6c0d.png)

schaferjh commented:
All of this is looking great to me. :fireworks: 

One very small comment: For the **text introduction** only the heading field is absolutely required. That way we can use this element for pages that just have a title (Contact Us) or pages that have a title, and an intro and some body copy and a link. From what I can tell at the moment only the link is optional -- but I may be misreading things!

KimberlyMunoz commented:
It looks like the Intro Paragraph and Title are marked as required, the body paragraph and link aren't required.

kave commented:
Yup i can make the change

kave commented:
@KimberlyMunoz Fixed the extra block space with if cases and fixed well for landing page (properly this time lol) and updated the required fields for text intro

jimmynotjim commented:
Justin just updated the Intro Specs to requiring the title and body text, but not the intro or link. We should nail this one down so we don't have to keep adjusting it.

duelj commented:
Sorry thing I misspoke, I would follow what Jess has said. My apologies.

kave commented:
PSA: ignore all the migrations scripts i'll be merging them into 1 file at the end of this PR

KimberlyMunoz commented:
![giphy](https://cloud.githubusercontent.com/assets/1860176/11307406/eb1d02e4-8f85-11e5-8344-d22fdc749312.gif)

:+1:

kave commented:
:dancer:
